packages/api/internal/template-manager/template_manager.go

```diff
@@ -1,20 +1,177 @@
 package template_manager
 
+import (
+	"context"
+	"github.com/e2b-dev/infra/packages/api/internal/utils"
+	"github.com/e2b-dev/infra/packages/shared/pkg/db"
+	"github.com/e2b-dev/infra/packages/shared/pkg/grpc/template-manager"
+	"github.com/e2b-dev/infra/packages/shared/pkg/models/envbuild"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+	"io"
+	"sync"
+	"time"
+)
+
+type processingBuilds struct {
+	templateID string
+}
+
 type TemplateManager struct {
-	grpc *GRPCClient
+	grpc       *GRPCClient
+	db         *db.DB
+	lock       sync.Mutex
+	processing map[uuid.UUID]processingBuilds
 }
 
-func New() (*TemplateManager, error) {
+var (
+	syncInterval             = time.Minute * 1
+	syncTimeout              = time.Minute * 5
+	syncWaitingStateDeadline = time.Minute * 20
+)
+
+func New(db *db.DB) (*TemplateManager, error) {
 	client, err := NewClient()
 	if err != nil {
 		return nil, err
 	}
 
 	return &TemplateManager{
-		grpc: client,
+		grpc:       client,
+		db:         db,
+		lock:       sync.Mutex{},
+		processing: make(map[uuid.UUID]processingBuilds),
 	}, nil
 }
 
 func (tm *TemplateManager) Close() error {
 	return tm.grpc.Close()
 }
+
+func (tm *TemplateManager) BuildsStatusPeriodicalSync(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(syncInterval):
+			buildsRunning, err := tm.db.GetRunningEnvBuilds(ctx)
+			if err != nil {
+				zap.L().Error("Error getting running builds for periodical sync", zap.Error(err))
+				time.Sleep(time.Second) // just not hammer database if there is an error
```

Comment:
How could it hammer database? Wouldn't it be waiting for `syncInterval`? Also is there a reason why to use time.After over ticker?

Could you also add timeout for database query?

Reply:
> How could it hammer database?

Its not possible with this type of calls. Before there was just inifinite `for` with sleep but i re-implement it to support context cancel so its no more needed.

> Could you also add timeout for database query?

Done. Not sure about right amount for explicit value for timeout here but 5s should work for now.